cache/cache_test.go

```diff
@@ -0,0 +1,318 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"go.etcd.io/etcd/client/pkg/v3/types"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	embed "go.etcd.io/etcd/server/v3/embed"
+)
+
+func TestPrefixWatchDeliversEvent(t *testing.T) {
+	client := startEtcd(t)
+
+	tests := []struct {
+		name    string
+		watchFn func(ctx context.Context, cache *Cache, prefix string) clientv3.WatchChan
+	}{
+		{
+			name: "client_watch",
+			watchFn: func(ctx context.Context, _ *Cache, prefix string) clientv3.WatchChan {
+				return client.Watch(ctx, prefix, clientv3.WithPrefix())
+			},
+		},
+		{
+			name: "cache_watch",
+			watchFn: func(ctx context.Context, cache *Cache, prefix string) clientv3.WatchChan {
+				return cache.Watch(ctx, prefix, clientv3.WithPrefix())
+			},
+		},
+		// TODO: more test case like "cache_watch_single_key" when we support more features
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			ctx, cancel := context.WithTimeout(t.Context(), time.Second)
+			defer cancel()
+
+			testPrefix := fmt.Sprintf("hello/%s/", tt.name)
+
+			cache, err := New(client, testPrefix, WithHistoryWindowSize(32))
+			if err != nil {
+				t.Fatalf("New(...): %v", err)
+			}
+			t.Cleanup(cache.Close)
+			if err := cache.WaitReady(ctx); err != nil {
+				t.Fatalf("cache not ready: %v", err)
+			}
+
+			ch := tt.watchFn(ctx, cache, testPrefix)
+
+			if _, err := client.Put(ctx, fmt.Sprintf("%sworld", testPrefix), "42"); err != nil {
+				t.Fatalf("Put: %v", err)
+			}
+
+			event := waitEvent(t, ch)
+			if diff := cmp.Diff(fmt.Sprintf("%sworld", testPrefix), string(event.Kv.Key)); diff != "" {
+				t.Fatalf("unexpected key (-want +got):\n%s", diff)
+			}
+
+			quietTimer := time.NewTimer(100 * time.Millisecond)
+			defer quietTimer.Stop()
+			for {
+				select {
+				case resp := <-ch:
+					if len(resp.Events) != 0 {
+						t.Fatalf("unexpected extra events: %+v", resp.Events)
+					}
+				case <-quietTimer.C:
+					return
+				}
+			}
+		})
+	}
+}
+
+func TestLaggingWatcher(t *testing.T) {
+	const prefix = "/test/"
+	cli := startEtcd(t)
+
+	tests := []struct {
+		name       string
+		window     int
+		produce    int
+		wantEvents int // -1 == expect none; 0 == expect compacted gap
+		wantClosed bool
+	}{
+		// within-window: expect exactly produce events, channel stays open
+		{
+			name:       "fits_entirely",
```

Comment:
```suggestion
		// within-window: expect exactly produce events, channel stays open
		{
			name:       "Window larger than events produced, expect all events and channel open",
```
Don't use comments to explain tests.